### PR TITLE
[ENH] Add IGCI estimator for bivariate causal discovery

### LINKED ISSUE #3427: Add IGCI pairwise causal discovery estimator
## 1. Is your feature request related to a problem? Please describe.

pgmpy does not currently provide pairwise causal discovery estimators. ANM covers nonlinear additive-noise settings, but it is not designed for near-deterministic low-noise relationships where causal asymmetry is better captured by the independence between the cause distribution and the causal mechanism.

Information-Geometric Causal Inference is a complementary pairwise method for these regimes.

## 2. Additional context

This issue proposes adding an `IGCI` estimator under `pgmpy.causal_discovery`.

Expected behavior:

- subclass `_BaseCausalDiscovery` directly
- accept exactly two-column continuous data in `fit`
- implement slope-based scoring by default
- optionally support entropy-based scoring
- support configurable reference measures, such as `"uniform"` and `"gaussian"`
- handle tied observations defensively to avoid finite-difference instability
- populate `causal_graph_`, `adjacency_matrix_`, and `direction_score_` after fitting
- define `direction_score_` as `S(Y -> X) - S(X -> Y)`, where positive means the forward column order wins
- mark categorical data unsupported via sklearn tags

Suggested files:

- `pgmpy/causal_discovery/IGCI.py`
- `pgmpy/tests/test_causal_discovery/test_IGCI.py`
- update `pgmpy/causal_discovery/__init__.py`

Testing should include near-deterministic low-noise synthetic data, tied observations, constant inputs, NaN/inf handling, insufficient samples, both reference measures, and reproducibility under fixed `seed`.

## 3. Paper

Janzing et al. Information-geometric approach to inferring causal directions.

## 4. Reference codebase

- causal-learn or comparable IGCI reference implementations
- Mooij et al. benchmark implementations for Cause-Effect Pairs

## 5. Related

- Pairwise causal discovery enhancement proposal
- ANM estimator issue
- Pairwise benchmarking notebook issue

### PR DESCRIPTION
**Pull request by hanara2112**
**The following checklist is mandatory**

Your PR will be closed if you remove the checklist. Use of LLMs is **strictly forbidden** for any part of this checklist (including for improving language), and will result in a **ban** if we find any use of LLMs.

### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md)?
- [x] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft.
- [x] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

If you have used AI/LLMs for any assistance, please answer the following questions. Please refer [#2622](https://github.com/pgmpy/pgmpy/pull/2622) for an example of the level of detail we expect:

- [x] Have you reviewed our [AI usage policy](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md#ai-usage-policy)?
- [x] Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.

- Please list the AI tool(s) used, along with the model and its version used.

codex 5.6 sol

- Please describe in **detail** how and what you used AI assistance for? Please outline your whole workflow with the AI tool.

I used AI as a research and coding assistant rather than to generate the final implementation. It helped me understand the IGCI papers, compare the algorithm with existing pgmpy estimators, draft the initial implementation and tests, and build a benchmarking script for the Tübingen dataset. I independently reviewed and revised the mathematical formulas, API design, documentation, error handling, and tests to ensure they matched the original papers and the pgmpy codebase, using the benchmark only to validate correctness rather than optimize performance.

- What steps have you taken to verify that the changes correctly address the issue? What edge cases have you considered? Other than running tests, what else have you verified?

I tested the implementation using pytest, synthetic datasets with known causal directions, and edge cases such as invalid inputs, missing values, and tied observations. I also compared the original spacing-based entropy estimator with the optional SciPy methods and validated the implementation on all 108 Tübingen cause-effect pairs, where the best configuration achieved 61.1% accuracy (66/108), consistent with the expected performance of IGCI.

- Have you used AI for generating tests? Can you compress them into a smaller number of tests without losing coverage?

Yes. The assistant helped draft the initial test file. I then simplified and reorganized the tests by combining similar cases, using parameterized tests for invalid inputs, and reducing duplicate checks. The final test suite contains a set of focused tests that cover initialization, slope and entropy scoring, reverse causal direction, tied values, and invalid inputs.

### Issue number(s) that this pull request fixes
- Fixes #3427 

### List of changes to the codebase in this pull request
- Added IGCI.py implementing the bivariate IGCI estimator with slope and entropy scoring methods
- Exported IGCI through pgmpy/causal_discovery/__init__.py.
- Added a test suite (test_IGCI.py) covering functionality and edge cases.
- Updated the documentation with citations to Janzing (2012) and Mooij et al. (2016).

### CODE REVIEW

File: pgmpy/causal_discovery/IGCI.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    >>> igci.causal_graph_.edges()
+    OutEdgeView([('X', 'Y')])
+    >>> igci.forward_score_.round(5)
+    np.float64(0.22245)
+    >>> igci.backward_score_.round(5)
+    np.float64(1.66886)
+
+    References
+    ----------
+    - :cite:p:`mooij_2016`
```
Comment (ankurankan) [inline pgmpy/causal_discovery/IGCI.py:79]: This needs to be changed. We have a new way to specify this.

**Comment by hanara2112** [comment]
igci_figures

I benchmarked the implementation on all 108 Tübingen cause–effect pairs, where the true causal direction is already known. For each pair, I tested different combinations of preprocessing (uniform and Gaussian) and scoring methods (slope and entropy), forcing a prediction every time and measuring the overall accuracy.

The best results came from uniform preprocessing, with both the slope score and the original spacing entropy estimator achieving 61.1% accuracy (66/108). Gaussian preprocessing performed slightly worse, while SciPy's entropy estimators achieved only 18–37% accuracy, showing that they are poor substitutes for the original IGCI entropy estimator. These results are broadly consistent with Mooij et al. (2016), where IGCI also performed only moderately better than chance, confirming that the implementation behaves as expected.

**Comment by hanara2112** [comment]
Clarification on the Algorithm:
IGCI estimator works on two continuous variables at a time. For each pair, I first preprocess the data using either uniform scaling or gaussian scaling. I then score both possible causal directions (X->Y and Y->X) and choose the direction with the lower score. The implementation supports two scoring methods: the **slope score**, which measures how the effect changes as the cause changes, and the **entropy score**, which compares the marginal entropy of the two variables. By default, the entropy method uses the original spacing-based estimator proposed by Mooij et al., while SciPy's differential entropy estimators are included as optional alternatives.

File: pgmpy/causal_discovery/IGCI.py (hunk 1 above)
Comment (hanara2112) [reply pgmpy/causal_discovery/IGCI.py:79]: How exactly can i specify it?

File: pgmpy/causal_discovery/bivariate_scores.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+            return scores[0]()
+        raise ValueError(f"Unknown {algorithm.upper()} score: {score!r}.")
+
+    if callable(score) and not isinstance(score, type):
+        return score
+
+    raise ValueError(f"Invalid {algorithm.upper()} score: {score!r}. Pass a built-in name or callable.")
+
+
+def get_anm_score(score: str | Callable) -> Callable:
```
Comment (ankurankan) [inline pgmpy/causal_discovery/bivariate_scores.py]: Just have a single method call `get_score`. Once that is done, remove the `_resolve_score` method and add its logic in the `get_score` method itself.

File: pgmpy/causal_discovery/anm_scores.py
Comment (ankurankan) [inline pgmpy/causal_discovery/anm_scores.py:1]: Remove this file. We do not need backward compatibility here as this is an unreleased feature

File: pgmpy/causal_discovery/bivariate_scores.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+    """
+    Mean log-slope score for IGCI.
+
+    This implements Equation 19 from the IGCI method. Repeated cause or effect values are not
+    supported; use :class:`WeightedSlopeScore` when the cause contains repetitions.
+    """
+
+    _tags = {"name": "slope", "supported_algorithms": ["igci"]}
+
+    def __call__(self, cause, effect) -> float:
```
Comment (ankurankan) [inline pgmpy/causal_discovery/bivariate_scores.py]: This needs to be `x` and `y`. Similar to the ANM scores. `cause` and `effect` also imply asymmetry in the score which is not the case.

File: pgmpy/causal_discovery/bivariate_scores.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    """
+    Repetition-aware log-slope score for IGCI.
+
+    This implements Equation 21 from the IGCI method and weights distinct cause values by their
+    original multiplicities.
+    """
+
+    _tags = {"name": "slope_weighted", "supported_algorithms": ["igci"]}
+
+    def __call__(self, cause, effect) -> float:
```
Comment (ankurankan) [inline pgmpy/causal_discovery/bivariate_scores.py]: Same here.

File: pgmpy/causal_discovery/bivariate_scores.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+        Returns
+        -------
+        float
+            Score for the two samples.
+        """
+        raise NotImplementedError
+
+
+# Backward-compatible name retained for the public ANM score API.
+BaseANMScore = BaseBivariateScore
```
Comment (ankurankan) [inline pgmpy/causal_discovery/bivariate_scores.py]: Can have just `BaseBivariateScore`.

File: pgmpy/causal_discovery/bivariate_scores.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+            score = test.effect_size_
+        elif self.criterion == "statistic":
+            score = abs(test.statistic_)
+        elif self.criterion == "p_value":
+            score = -test.p_value_
+        else:
+            raise ValueError(
+                f"Unknown criterion: {self.criterion!r}. Must be one of 'effect_size', 'statistic', 'p_value'."
+            )
+        return _ensure_finite(score)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/bivariate_scores.py]: I don't think we need this. Under what scenario would we get a non-finite score here?

File: pgmpy/causal_discovery/bivariate_scores.py (hunk 2 above)
Comment (hanara2112) [reply pgmpy/causal_discovery/bivariate_scores.py]: done

File: pgmpy/causal_discovery/bivariate_scores.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+        run_start = np.r_[0, np.flatnonzero(np.diff(x)) + 1]
+        multiplicities = np.diff(np.r_[run_start, x.size])
+        x = x[run_start]
+        y = y[run_start]
+
+        if x.size < 2:
+            raise ValueError(
+                "SlopeScore requires sufficiently distinct x values after removing repetitions; "
+                "no valid pairs remained."
+            )
```
Comment (ankurankan) [inline pgmpy/causal_discovery/bivariate_scores.py]: Checks should always be towards the top of the method if possible.

File: pgmpy/causal_discovery/bivariate_scores.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+
+        x_diff = np.diff(x)
+        y_diff = np.diff(y)
+        valid = y_diff != 0
+        if not valid.any():
+            raise ValueError("SlopeScore requires at least one non-zero y spacing.")
+
+        score = np.average(
+            np.log(np.abs(y_diff[valid] / x_diff[valid])),
```
Comment (ankurankan) [inline pgmpy/causal_discovery/bivariate_scores.py:228]: We will need to validate `x_diff` as well.

**Review by ankurankan — CHANGES_REQUESTED** [review]

File: pgmpy/causal_discovery/bivariate_scores.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        raise ValueError("Entropy estimation requires at least two observations.")
+
+    deltas = np.diff(values)
+    if np.any(deltas == 0):
+        raise ValueError("Spacing entropy requires distinct observations.")
+
+    entropy = psi(values.size) - psi(1) + np.log(deltas).sum() / (values.size - 1)
+    if base is not None:
+        entropy /= np.log(base)
+    return _ensure_finite(entropy, name="Entropy estimate")
```
Comment (ankurankan) [inline pgmpy/causal_discovery/bivariate_scores.py]: Is there a way to specify arguments in `differential_entropy` such that it computes this spacing entropy?

**Review by ankurankan — APPROVED** [review]
